### ISSUE TITLE
$repo_url not working

### ISSUE THREAD

**Issue by khaledosman** [juliuscc/semantic-release-slack-bot#24]
The plugin sends the message to slack literally as `$repo_url` instead of the actual repository url

package.json
```
{
    ...
  "repository": {
    "type": "git",
    "url": "<repoUrl>"
  },
  "config": {
    "commitizen": {
      "path": "./node_modules/cz-conventional-changelog"
    }
  },
  "release": {
    "plugins": [
      "@semantic-release/commit-analyzer",
      "@semantic-release/release-notes-generator",
      [
        "semantic-release-slack-bot",
        {
          "notifyOnSuccess": true,
          "notifyOnFail": false,
          "onSuccessTemplate": {
            "text": "A new version of $package_name with version $npm_package_version has been released at $repo_url ! \n $release_notes"
          },
          "markdownReleaseNotes": true
        }
      ],
      [
        "@semantic-release/npm",
        {
          "npmPublish": false
        }
      ],
      "@semantic-release/changelog",
      "@semantic-release/git"
    ]
  },
}
```

**Comment by juliuscc** [comment]
Thanks for the issue! That is very strange. Do you know if the other variables (package_name, npm_package_version, ...) are working?

**Comment by khaledosman** [comment]
yes, these work fine
here's an example message generated by the above configuration
![image](https://user-images.githubusercontent.com/2937633/69224052-21414f80-0b7c-11ea-96fa-d4cad68f8ee1.png)

**Comment by juliuscc** [comment]
Okay, then we now that the templater is working, but is not getting all variables for some reason.

1. Are you running semantic-release as an npm-script or with npx?
2. What happens if you change the repository field to only the string instead of an object?

**Comment by juliuscc** [comment]
Is it possible for you to give me the repository object from your package.json? This way I could try stuff myself.

**Comment by khaledosman** [comment]
I'm running semantic-release as an npm-script that runs npx as follows
```
"scripts": {
    ...
    "release": "npx semantic-release",
    ...
}
```
and here's the repository object
```
"repository": {
    "type": "git",
    "url": "git+ssh://git@bitbucket.org/colorfy/ovs-data-pusher.git"
  },
```

**Comment by BeyondEvil** [comment]
tl;dr: `repoUrl` is undefined.

Without having tested this, here's my guess about what's happening.

The repository url is of the `bitbucket` variety. Which means that [`repoPath`](https://github.com/juliuscc/semantic-release-slack-bot/blob/master/lib/success.js#L28) will be `undefined.`

And since `repoPath` is undefined, `repoUrl` will be undefined.

The way the templater works, is that if a replacement string isn't present (undefined), we just add it verbatim. See this test: https://github.com/juliuscc/semantic-release-slack-bot/blob/master/test/test_template.js#L12

**Comment by juliuscc** [comment]
Thanks @BeyondEvil!

Does this mean that we only parse out repoURL if it is a link from github? When I looked at that piece of code earlier today I thought it meant we nicely formated github links but just passed through other links.

**Comment by juliuscc** [comment]
I will look more into implementing a general solution @khaledosman, but it might take some while (1-2 weeks) as I am pretty swamped at the moment. If you would like to give it a try, you are very welcome to submit a PR 🙂

**Comment by BeyondEvil** [comment]
Basically if `git@github.com` is not present in the repository url, we set `repoPath` to undefined.
And like mentioned, if `repoPath` is undefined, `repoURL` is undefined.

However, if the repo url _does_ contain `git@github.com` `repoURL` will be turned into the https equivalent.

**Comment by juliuscc** [comment]
Thanks for offering help!

My initial thought was to transform any ssh-git link into a link to the repo, but that might not be possible as different vendors might have different kinds of URLs to the repo, or might not even have a repo web page.

I think that most large vendors should be supported, like GitHub, Gitlab, Bitbucket, and Azure (maybe more if I missed some large one), instead of only supporting GitHub. It should also say in the description which vendors are supported.

Basically, I would like to have that `repoPath` and `repoUrl` worked for more vendors than only GitHub.

**Comment by BeyondEvil** [comment]
Actually, they have to follow the git [protocol](https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols)

I'll create a PR with a suggested solution. It will at least work for vendors that follow the protocol.

**Comment by juliuscc** [comment]
Yes, for the repo_url but the URL for viewing a repo online is not standardized. A vendor doesn't even have to have a web page for the remote server.

**Comment by BeyondEvil** [comment]
Ok, here's what I've found:

`GitHub`: `https://github.com/<user>/<project>`

`GitLab`: `https://gitlab.com/<user>/<project>`

`Bitbucket`: `https://bitbucket.org/<user>/<project>` which then redirects to `https://bitbucket.org/<user>/<project>/src/default/`

Couldn't figure `Azure` out 🤷‍♂ 

So the pattern does seem to be: `https://<vendor domain>/<user>/<project>`

I'll create a PR supporting that, and if there's vendors out there that we need to support we can extend the functionality. Regardless, it will be an improvement. 😊

**Comment by BeyondEvil** [comment]
Sorry to bother you @juliuscc but what are the requirements for `repoPath`?

Also, see my previous post. 😊

**Comment by juliuscc** [comment]
Great job! I think this will be enough. We could write that this is tested for GitHub, gitlab, and bitbucket in the readme.

> what are the requirements for repoPath?

What do you mean with this question

**Comment by BeyondEvil** [comment]
> What do you mean with this question

Well, given `https://github.com/some-user/some-project`, what will the path be? The entire thing, just `some-user/some-project`, or something else? 😊 

@juliuscc

**Comment by juliuscc** [comment]
Ah! `repoPath` as it is now is only `some-user/some-project` but I am not sure if it should be exposed at all. I do not see the use case for it. Especially when there exists `repoUrl`. What do you think?

**Comment by BeyondEvil** [comment]
I'm too new to the semantic-release domain to tell what potential use cases may be (or are!).

I can do some local testing to see what the impact could be. Afaict from the code, the only place it's really used is as an attachment to the slack message.

The worst that can happen is that we make a change that we later have to revert. I don't think we'll ruin anyone's life. 😊

**Comment by BeyondEvil** [comment]
Sorry for the delay! PR added.

However, I have not had time to review whether or not `repo_path` should live on. So this PR shouldn't introduce any breaking changes.

**Comment by juliuscc** [comment]
Yeah there is no harm in keeping it, and beeing backwards compatible. Thanks for the great work!
